### PR TITLE
[Sanitizers] Add Scudo support

### PR DESCRIPTION
**Pull request by Lukasa**
LLVM ships a hardened memory allocator called Scudo:
https://llvm.org/docs/ScudoHardenedAllocator.html. This allocator
provides additional mitigations against heap-based vulnerabilities, but
retains sufficient performance to be safely run in production
applications.

While ideal Swift applications are obviously written in pure Swift, in
practice most applications contain some amount of code written in
less-safe languages. Additionally, plenty of Swift programs themselves
contain unsafe code, particularly when attempting to implement
high-performance data structures. These sources of unsafety introduce
the risk of memory issues, and having the option to use the Scudo
allocator is a useful defense-in-depth tool.

This patch enables `-sanitize=scudo` as an extra `swiftc` flag. This
sanitizer is only supported on Linux, so no further work is required to
enable it on Windows or Apple platforms. As this "sanitizer" is only a
runtime component, we do not require any wider changes to instrument
code.

The result of this patch matches the API of clang, which supports
`-fsanitize=scudo`.

- - - 

I do have one question: currently this patch contains no regression tests. I was fundamentally unsure of where to place these regression tests, but I think we should have them, so please do as part of your review suggest a location and form for the regression tests.

### CODE REVIEW

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
I should note that, when we update the version of llvm we're using, enabling Scudo will also enable [GWP-Asan](https://llvm.org/docs/GwpAsan.html), which provides not just mitigation but active detection of problems.

**Comment by danliew-apple** [comment]
@Lukasa Thanks for the patch. There are something things that need to be addressed before this can land.

1. Given that this only works on Linux you need to add code in the Swift driver to refuse to use `-sanitize=scudo` on non Linux platforms. I can't see anything obvious in the code you added that forbids this.
2. This patch doesn't change anything about the compilation process. AFAIK Scudo is a replacement allocator library so I would expect that the linker would need to be told to link against the scudo runtime.
3. This is missing tests. You need:

* A test for successful driver invocation on Linux. You should probably check that the linker flags look right (i.e. links against Scudo runtime).
* A test for driver failure on non-Linux platforms.
* Tests for the invalid santizer+scudo combination.
* At least one test case where we compile and run a program that uses Scudo.

**Review by ghost — CHANGES_REQUESTED** [review]
See my initial comments.

**Comment by Lukasa** [comment]
Thanks @danliew-apple!

> Given that this only works on Linux you need to add code in the Swift driver to refuse to use -sanitize=scudo on non Linux platforms. I can't see anything obvious in the code you added that forbids this.

I don't believe this is true. There are separate Sanitizer whitelists for [Darwin](https://github.com/apple/swift/blob/3ff6c1c315677789cc96a84c9e17ab9f5105614d/lib/Driver/DarwinToolChains.cpp#L304) and [Windows](https://github.com/apple/swift/blob/238d097af906ea59a7343142d5b28862b2bfc97a/lib/Driver/WindowsToolChains.cpp#L162) which explicitly whitelist the sanitisers they support. I have observed that with this toolchain built on Darwin the Scudo flag is rejected: I believe this is true for Windows as well, though @compnerd could further clarify.

> This patch doesn't change anything about the compilation process. AFAIK Scudo is a replacement allocator library so I would expect that the linker would need to be told to link against the scudo runtime.

It turns out it already is. On [Linux](https://github.com/apple/swift/blob/eae8d028ba469e11941dc6939b34fbb26c7d5fea/lib/Driver/UnixToolChains.cpp#L291-L301) the driver will simply pass the sanitizer option through to clang, which drives the actual linking process. This results in the sanitising being applied. Again, I've verified this in live code.

> This is missing tests.

Fantastic. Can you point me to places in the code that do similar testing? I wasn't able to find them, and that made it very hard to write more.

**Comment by danliew-apple** [comment]
@Lukasa Here are some examples:

* [Here's an example driver test that checks that the compiler refuses certain flags](https://github.com/apple/swift/blob/master/test/Driver/sanitize_recover.swift). The test is also checking that the driver passes an option to the Swift frontend. That isn't relevant in this case so you can ignore that.
* [This test](https://github.com/apple/swift/blob/master/test/Driver/sanitizers.swift) does a bunch of things but one the things it does is check that a sanitizer runtime library is being passed to the linker.
* [Here's an example of a test that builds and executes code](https://github.com/apple/swift/blob/master/test/Sanitizers/asan_recover.swift)

Note that `llvm-lit` has a notion of features and you can use this to make a test run on only certain platforms. My suggestion is that

* Tests for driver options should probably execute on all platforms. In these tests you'd explicit specify the target so that we can verify that the driver allows Scudo on Linux but rejects it for Windows on macOS.
* Tests that are only going to work on Linux should be annotated with `REQUIRES: OS=linux-gnu`.

**Comment by Lukasa** [comment]
Thanks @danliew-apple, I’ll write some tests to cover this and ping for re-review when I’ve done so.

**Comment by danliew-apple** [comment]
> I don't believe this is true. There are separate Sanitizer whitelists for Darwin and Windows which explicitly whitelist the sanitisers they support. I have observed that with this toolchain built on Darwin the Scudo flag is rejected: I believe this is true for Windows as well, though @compnerd could further clarify.

The code you link to is related to linking. The **driver** should fail compilation much earlier than that.  If someone passes `-sanitizer=scudo` option for any more of the driver and the platform doesn't support it then it should fail.

> This patch doesn't change anything about the compilation process. AFAIK Scudo is a replacement allocator library so I would expect that the linker would need to be told to link against the scudo runtime.
>> It turns out it already is. On Linux the driver will simply pass the sanitizer option through to clang, which drives the actual linking process. This results in the sanitising being applied. Again, I've verified this in live code.

Interesting. I would not have expected that given that the Swift compiler shouldn't know what to do with this new enum you've added. You'll need to show me working tests to convince me.

**Comment by danliew-apple** [comment]
@Lukasa Just to be clear I'm only giving you a review the code as I see it. I don't work on the swift compiler very frequently so I think you will need approval from others to land this. Given that this is very untested I suspect we might want to label it as experimental.

**Comment by Lukasa** [comment]
@danliew-apple Totally understood, and I’m very appreciative of your time. I’m confident the patch-as-written works in the straightforward cases because I tested it live, but I also totally agree that without the tests it is hard for anyone else to validate my confidence! So I absolutely appreciate the feedback. I’ll work on adding those tests, and we can worry about wider approval once I’ve dealt with this feedback.

**Comment by danliew-apple** [comment]
@airspeedswift @mikeash Does adopting the hardened Scudo on Linux seem a reasonable thing to support?

**Comment by danliew-apple** [comment]
@Lukasa I just noticed that the Scudo documentation says.. "Compiling with Scudo will also enforce PIE for the output binary". That's the behaviour for Clang. Do you know why they do this? If there's a good reason for this we might need to do the same thing for Swift. I don't know if Swift has an equivalent PIE mode.

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by mikeash** [comment]
@danliew-apple I'm not too familiar with it, but reading through the link you provided, this seems like a great thing to support. Especially since the hard work is in LLVM and we apparently just need to support the option and pass it along. @weissi Any additional thoughts?

**Comment by weissi** [comment]
@mikeash / @Lukasa this looks good to me, IMHO definitely something we want :)

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please test

**Comment by Lukasa** [comment]
Solid work Jenkins, thanks for that.

**Comment by Lukasa** [comment]
@danliew-apple I believe I have covered all the bases with testing here. I have validated that the driver correctly rejects the Scudo "sanitizer" on all non-Linux platforms, and that Scudo is successfully enabled on Linux and works (by looking for a Scudo-specific error message on a double-free). Hopefully the test coverage that exists here will be sufficient to convince folks that the functionality does work as intended.

**Comment by Lukasa** [comment]
The OS X platform failure seems unrelated to this change, everything else is shiny and happy.

**Review by jckarter — APPROVED** [review]
LGTM, shouldn't affect anyone who isn't explicitly reaching for this flag.

File: lib/Option/SanitizerOptions.cpp
Comment (yln) [inline lib/Option/SanitizerOptions.cpp]: Note: this if-else needs to be extended whenever the SanitizerKind enum gets extended, but I don't see a better way of doing this with the current API of `OptionSet`.

Comment (Lukasa) [reply lib/Option/SanitizerOptions.cpp]: Yup. `OptionSet` absolutely could answer this question, but I don’t know how much value we get from ramming that new single-use API into this patch.

Comment (yln) [reply lib/Option/SanitizerOptions.cpp]: Not requesting anything additional for this PR, just wanted to point it out; maybe someone has an idea.

Comment (Lukasa) [reply lib/Option/SanitizerOptions.cpp]: Ah sorry, I understood you weren’t asking for new API, I was just adding clarification on why I didn’t add it as part of this patch.

File: test/Driver/sanitize_scudo.swift
Comment (danliew-apple) [inline test/Driver/sanitize_scudo.swift]: Are these on the same line? If so this should probably be `SCUDO_LINUX-SAME:`.

File: test/Driver/sanitize_scudo.swift
Comment (danliew-apple) [inline test/Driver/sanitize_scudo.swift]: I'd still like an answer on this. As I understand it using `-pie` requires that the objects files be position independent code but I have idea if the Swift compiler actually guarantees this.

Comment (jckarter) [reply test/Driver/sanitize_scudo.swift]: Swift should always use pie.

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
Hmm, the CI is very sad.

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
Thanks for the review @danliew-apple, I think I've addressed your feedback.

**Comment by compnerd** [comment]
Hmm, do we not want the runtime to have a scudo variant as well if we are going to support this?  Having two memory domains is usually a good recipe for problems (what happens if someone does a `UnsafeMemoryPointer.deallocate` from Swift on something allocated from C with scudo intercepting the allocation?)

**Comment by Lukasa** [comment]
Nope, there is one and only one allocator in a program at any one time. If scudo is on, it is on for everything, because it is linked into the main binary. In some senses this is a strength over ASan: as it operates at the allocator level there is no reason or need to provide customised runtimes or sanitizer hooks.

I encourage you to try it out: write a Swift package that allocates in C, frees in C, and is then deallocated again in Swift (for a double free). Scudo will catch it.

**Comment by Lukasa** [comment]
Hey @danliew-apple, do you have any chance of looking at this this week?

**Comment by Lukasa** [comment]
A gentle ping to @danliew-apple.

File: test/Driver/sanitize_scudo.swift
Comment (danliew-apple) [inline test/Driver/sanitize_scudo.swift]: This target triple is incorrect. It should be `i386-apple-macosx10.9`.

File: test/Driver/sanitize_scudo.swift
Comment (danliew-apple) [inline test/Driver/sanitize_scudo.swift]: This `--implicit-check-not=SCUDO_LINUX` line doesn't look right. That's adding checks that we don't see `SCUDO_LINUX` in the compiler's output which doesn't make sense.

I presume you meant `-check-prefix=SCUDO_LINUX --implicit-check-not=<some pattern>`?

Where `<some pattern>` is whatever you don't what to appear in the compiler's output.

File: test/Driver/sanitize_scudo.swift
Comment (danliew-apple) [inline test/Driver/sanitize_scudo.swift]: You need to update the target triple here too.

**Review by ghost — CHANGES_REQUESTED** [review]
Most of this looks good. The tests need a few fixes though.

I'd also like you to change the commit message to explain why using Scudo gets rejected for other platforms because it's really unobvious looking at the patch. I had to compile and run the code to figure out what was going on. Maybe something like this:


_The Swift driver rejects platforms that don't support Scudo using an existing mechanism in the Driver that is not part of this patch. This mechanism is in `swift::parseSanitizerArgValues(...)` (`lib/Option/SanitizerOptions.cpp`). The mechanism determines if a sanitizer is supported by checking for the existence of the corresponding sanitizer runtime library in the compiler's resource directory. The Scudo runtime library currently only exists in the Linux compiler resource directory. This results in the driver only allowing Scudo when targeting Linux._

**Comment by danliew-apple** [comment]
@Lukasa Sorry for the delay in reviewing this.

File: test/Driver/sanitize_scudo.swift
Comment (Lukasa) [reply test/Driver/sanitize_scudo.swift]: Really? TSAN uses this one as well, we should consider fixing that up.

File: test/Driver/sanitize_scudo.swift
Comment (Lukasa) [reply test/Driver/sanitize_scudo.swift]: I think this is right: when building only libraries, we don't want to add the sanitiser because it should only be added when building the main binary. This is the same as the requirement on all the other sanitisers with runtime libraries: only the main binary actually invokes the linker to link the sanitiser runtime.

**Comment by Lukasa** [comment]
Thanks @danliew-apple, I've responded to one or two of your points where I think the patch is currently correct, and have addressed your others, including updating the commit message. Do you mind taking another look?

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
@swift-ci please test

File: test/Driver/sanitize_scudo.swift
Comment (danliew-apple) [reply test/Driver/sanitize_scudo.swift]: Yes. The Swift compiler rejects the `x86-apple-macosx10.9` target triple. And yes if you copy and pasted this from a TSan test we should fix that too but in a separate patch.

**Review by ghost — CHANGES_REQUESTED** [review]
There's still an issue with the test that needs to be resolved.

I also have a very minor nit with the commit message.


_The result of this patch matches the API of clang, which supports
`-fsanitize=scudo`._


Technically `-fsanitize=scudo` isn't really an API and even then its not an exact match because the command line flag name is different.

Maybe when you say

_This patch enables `-sanitize=scudo` as an extra `swiftc` flag._

You could say as the next sentence.

_This is similar to Clang's `-fsanitize=scudo` flag_.

**Comment by Lukasa** [comment]
@swift-ci please smoke test

**Comment by Lukasa** [comment]
Thanks for that clarification @danliew-apple, that helped a lot! I have rewritten that test and also updated the commit message, please take another look.

**Comment by Lukasa** [comment]
@swift-ci please test

**Review by ghost — APPROVED** [review]

**Comment by danliew-apple** [comment]
@swift-ci please test

**Comment by danliew-apple** [comment]
@Lukasa Sorry for being slow. I've approved the changes. Thanks for being patient.

**Comment by Lukasa** [comment]
No need to apologise, I understand completely that we’re all busy! Thanks so much for your time on this.

**Comment by danliew-apple** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
I believe that this has caused a regression on Windows: https://ci-external.swift.org/job/oss-swift-windows-x86_64/2607/

**Comment by Lukasa** [comment]
Good catch @compnerd, that fix looks right. Sorry about that!

**Pull request merged**
